### PR TITLE
[docs] Revised Cluster documentation

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?
The Ray Autoscaler/multi-node documentation needed revising.

This makes it easier for users to find information on how to launch ray clusters and deploy ray.

QUESTION: This also renames "ray autoscaler" to "ray cluster launcher". Is this OK?

image

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: doc/source/debugging.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
+Debugging
+=========
+
+Starting processes in a debugger
```
Comment (anabranch) [inline doc/source/debugging.rst:4]: This seems unrelated, is this the wrong PR?

File: doc/source/getting-involved.rst [hunk 2]
```diff
@@ -1,5 +1,17 @@
-Getting Involved
-================
+Getting Involved / Contributing
```
Comment (anabranch) [inline doc/source/getting-involved.rst:1]: wrong PR..

File: doc/source/cluster-index.rst [hunk 3]
```diff
@@ -1,15 +1,105 @@
-Deploying Ray
-=============
+Multi-node Ray (Overview)
```
Comment (anabranch) [inline doc/source/cluster-index.rst:1]: first things first, we want to structure this so that all files related to the autoscaler are under `/cluster`. For instance:
-  `doc/source/cluster/index.rst`
-  `doc/source/cluster/configuration.rst`
-  `doc/source/cluster/cloud.rst`
-  `doc/source/cluster/cluster-managers.rst`

With the structure, it makes it very easy for people to navigate the tree and know exactly where something should go. By removing from top level you also make contributions easier ("just throw it in the cluster folder").

File: doc/source/cluster-launcher-usage.rst [hunk 4]
```diff
@@ -0,0 +1,193 @@
+.. _launcher-usage:
+
+Launcher Usage Guide
```
Comment (anabranch) [inline doc/source/cluster-launcher-usage.rst:3]: A "Launcher" is a bit imprecise. Can we call this "Launching Clusters" or something?

File: doc/source/cluster-index.rst (hunk 3 above)
Comment (anabranch) [reply doc/source/cluster-index.rst:1]: Index should just be what this file is.

**Review by bllchmbrs — CHANGES_REQUESTED** [review]

File: doc/source/cluster-launcher-usage.rst [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+See the :ref:`Cluster Configuration <cluster-config>` docs on how to customize the configuration file.
+
+Here are the commands covered in this document:
+
+.. contents::
+    :local:
+    :backlinks: none
+
+Launching a cluster (``ray up``)
```
Comment (anabranch) [inline doc/source/cluster-launcher-usage.rst:43]: this structure is awesome btw!

File: doc/source/cluster-launcher-usage.rst (hunk 4 above)
Comment (richardliaw) [reply doc/source/cluster-launcher-usage.rst]: it's a bit like a package reference though (covering main commands)

Comment (richardliaw) [reply doc/source/cluster-launcher-usage.rst]: renamed to 'cluster launcher usage guide'

File: doc/source/cluster/cloud.rst [hunk 6]
```diff
@@ -0,0 +1,225 @@
+.. _cluster-cloud:
+
+Launching clusters on the cloud
+===============================
+
+This section provides instructions for configuring the Ray Cluster Launcher to use with AWS/Azure/GCP, an existing Kubernetes cluster, or on a private cluster of host machines.
```
Comment (anabranch) [inline doc/source/cluster/cloud.rst:6]: I would recommend linking to my [blog post here](https://medium.com/distributed-computing-with-ray/a-step-by-step-guide-to-scaling-your-first-python-application-in-the-cloud-8761fe331ef1)

In general, it's a great way to provide links for people to discover other content.

File: doc/source/cluster/deploy.rst [hunk 7]
```diff
@@ -0,0 +1,16 @@
+.. _ref-cluster-setup:
+
+Ray with Cluster Managers
+=========================
+
+.. note::
+
+    If you're using AWS, Azure or GCP you should use the automated :ref:`setup commands <ref-automatic-cluster>`.
```
Comment (anabranch) [inline doc/source/cluster/deploy.rst:8]: why?

File: doc/source/cluster/kubernetes.rst [hunk 8]
```diff
@@ -1,18 +1,13 @@
...
-  documentation <autoscaling.html>`__ for details.
-
-.. warning::
-
-  Running Ray on Kubernetes is still a work in progress. If you have a
-  suggestion for how to improve this documentation or want to request a
-  missing feature, please get in touch using one of the channels in the 
-  `Questions or Issues?`_ section below.
+  which has support for running on top of Kubernetes. Please see the :ref:`autoscaler
+  documentation <ray-launch-k8s>` for details.
```
Comment (anabranch) [inline doc/source/cluster/kubernetes.rst:10]: What is the purpose of this document? You mention this reference to another doc but don't really say what the goal of this page will be.

File: doc/source/index.rst [hunk 9]
```diff
@@ -146,10 +146,17 @@ Getting Involved
    using-ray.rst
    configure.rst
    ray-dashboard.rst
-   cluster-index.rst
    Tutorial and Examples <auto_examples/overview.rst>
    package-ref.rst
 
+.. toctree::
```
Comment (anabranch) [inline doc/source/index.rst:152]: See how nice this looks :)

**Review by bllchmbrs — APPROVED** [review]
Minor changes requested, so it's basically LGTM pending those changes.

File: doc/examples/plot_example-lm.rst [hunk 10]
```diff
@@ -4,18 +4,19 @@ Fault-Tolerant Fairseq Training
 This document provides a walkthrough of adapting the `Fairseq library <https://github.com/pytorch/fairseq>`__ to perform fault-tolerant distributed training on AWS.
 As an example, we use the WikiText-103 dataset to pretrain the RoBERTa model following `this tutorial <https://github.com/pytorch/fairseq/blob/master/examples/roberta/README.pretraining.md>`__. The pipeline and configurations in this document will work for other models supported by Fairseq, such as sequence-to-sequence machine translation models.
 
-To run this example, you will need to install Ray on your local machine to use Ray Autoscaler.
+To run this example, you will need to install Ray on your local machine to use Ray cluster launcher.
```
Comment (edoakes) [inline doc/examples/plot_example-lm.rst:7]:
```suggestion
To run this example, you will need to install Ray on your local machine to use the Ray cluster launcher.
```

File: doc/examples/plot_example-lm.rst [hunk 11]
```diff
@@ -4,18 +4,19 @@ Fault-Tolerant Fairseq Training
...
 
 You can view the `code for this example`_.
 
 .. _`code for this example`: https://github.com/ray-project/ray/tree/master/doc/examples/lm
 
 
-To use Ray Autoscaler on AWS, install boto (``pip install boto3``) and configure your AWS credentials in ``~/.aws/credentials`` as described on  `Automatic Cluster Setup page <../autoscaling.html>`__. We provide an `example config file <https://github.com/ray-project/ray/tree/master/doc/examples/lm/lm-cluster.yaml>`__ (``lm-cluster.yaml``).
+To use Ray cluster launcher on AWS, install boto (``pip install boto3``) and configure your AWS credentials in ``~/.aws/credentials`` as described on  :ref:`Automatic Cluster Setup page <ref-automatic-cluster>`.
```
Comment (edoakes) [inline doc/examples/plot_example-lm.rst:14]:
```suggestion
To use Ray cluster launcher on AWS, install boto (``pip install boto3``) and configure your AWS credentials in ``~/.aws/credentials`` as described on the  :ref:`Automatic Cluster Setup page <ref-automatic-cluster>`.
```

File: doc/source/cluster/cloud.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (edoakes) [reply doc/source/cluster/cloud.rst:6]: Yeah this is a good point, users are pretty confused about having two ways to launch on k8s. We should mention here and on that page that by default they should use cluster launch but if they need additional features/manual control, they can deploy manually (that's what the other page is for)

File: doc/source/cluster/index.rst [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+Multi-node Ray Overview
+=======================
+
+One of Ray's strengths is the ability to leverage multiple machines in the same program. You can do this by creating and connecting to a Ray cluster.
+
+Key Concepts
+------------
+
+* **Ray Nodes**: A Ray cluster consists of a **head node** and set of **worker nodes**. The head node needs to be started first, and the worker nodes are given the address of the head node to form the cluster. The Ray cluster itself can also "auto-scale", meaning that it can interact with a Cloud Provider to request or release instances according to application workload.
```
Comment (edoakes) [inline doc/source/cluster/index.rst:11]:
```suggestion
* **Ray Nodes**: A Ray cluster consists of a **head node** and a set of **worker nodes**. The head node needs to be started first, and the worker nodes are given the address of the head node to form the cluster. The Ray cluster itself can also "auto-scale," meaning that it can interact with a Cloud Provider to request or release instances according to application workload.
```

File: doc/source/cluster/kubernetes.rst (hunk 8 above)
Comment (edoakes) [reply doc/source/cluster/kubernetes.rst:10]: +1, we should mention that this is for manual control/advanced usage. More users should use cluster launcher

**Review by edoakes — APPROVED** [review]
Looks fine to me. I'm positive about the autoscaler -> "cluster launcher" rename, but we should probably get more feedback from a wider audience.

File: doc/source/cluster/cloud.rst [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+.. contents::
+    :local:
+    :backlinks: none
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/cluster/cloud.rst:10]: https://docutils.sourceforge.io/docs/ref/rst/directives.html#table-of-contents

:local: is the main one that is important i think
